lib/internal/streams/add-abort-signal.js

```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+const {
+  AbortError,
+  codes,
+} = require('internal/errors');
+
+const eos = require('internal/streams/end-of-stream');
+const { ERR_INVALID_ARG_TYPE } = codes;
+
+const {
+  validateAbortSignal,
+} = require('internal/validators');
+
+function isStream(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+module.exports = function addAbortSignal(signal, stream) {
+  validateAbortSignal(signal, 'signal');
+  if (!isStream(stream)) {
+    throw new ERR_INVALID_ARG_TYPE('stream', 'stream.Stream', stream);
+  }
+  const onAbort = () => {
+    stream.destroy(new AbortError());
+  };
+  if (signal.aborted) {
+    onAbort();
```

Comment:
@mcollina does this need to be in nextTick?

Reply:
Note cancellation is intentionally synchronous in this case to behave like .destroy(new AbortError)

Why would we want to nextTick here?

Reply:
I’m not sure. Just making sure we’ve considered it.

Reply:
Not needed, events are nexticked already by destroy().